Java: Nullness library: track instanceof expressions in correlated conditions

Cornelius-Riemenschneider opened the pull request:
This fixes a FP on apache kafka, see #2238 .
@aschackmull

aschackmull commented on java/ql/src/semmle/code/java/dataflow/NullGuards.qll:
```diff
@@ -24,6 +24,33 @@ Expr enumConstEquality(Expr e, boolean polarity, EnumConstant c) {
+    or
+    result instanceof NEExpr and
+    isEqualExpr = false
+  )
```
This can be simplified by using `EqualityTest` and `EqualityTest.polarity()` instead of `BinaryExpr`.

aschackmull commented on java/ql/src/semmle/code/java/dataflow/NullGuards.qll:
```diff
@@ -24,6 +24,33 @@ Expr enumConstEquality(Expr e, boolean polarity, EnumConstant c) {
+}
+
+/**
+ * Gets an expression of the form `v1` == `v2` or `v1` != `v2`.
```
The `==` and the `!=` ought to be included in the `` ` ``s.

aschackmull commented on java/ql/src/semmle/code/java/dataflow/NullGuards.qll:
```diff
@@ -24,6 +24,33 @@ Expr enumConstEquality(Expr e, boolean polarity, EnumConstant c) {
+    or
+    result.getLeftOperand() = v2.getAUse() and
+    result.getRightOperand() = v1.getAUse()
+  ) and
```
The two cases can be combined into one by using `hasOperands`, which is already symmetric, see `enumConstEquality` above.

aschackmull commented on java/ql/src/semmle/code/java/dataflow/NullGuards.qll:
```diff
@@ -24,6 +24,33 @@ Expr enumConstEquality(Expr e, boolean polarity, EnumConstant c) {
 }
 
+/** Gets an instanceof expression of `v` with type `type` */
+InstanceOfExpr instanceofExpr(SsaVariable v, Expr type) {
```
It is better to move the `getType()` of the type expression into this predicate, as that allows a more direct join in the use of this predicate.

Cornelius-Riemenschneider commented:
Thanks for the review, I addressed everything and renamed varComparisonExpr to varEqualityTestExpr.
I don't have a FP for ==, only for != but I included it both for completeness and to handle the !(a == b) case.

aschackmull commented on java/ql/src/semmle/code/java/dataflow/NullGuards.qll:
```diff
@@ -25,30 +25,19 @@ Expr enumConstEquality(Expr e, boolean polarity, EnumConstant c) {
-  )
+EqualityTest varEqualityTestExpr(SsaVariable v1, SsaVariable v2, boolean isEqualExpr) {
+  result.hasOperands(v1.getAUse(), v2.getAUse()) and
+  result instanceof EqualityTest and
```
`result` is already specified as an `EqualityTest `, so this line is superfluous.

aschackmull commented:
The autoformat check on `java/ql/src/semmle/code/java/dataflow/Nullness.qll` failed.
